lib/core/versions/latest/DocumentComposer.ts

```diff
@@ -219,100 +221,136 @@ export default class DocumentComposer {
-        }
-        // NOTE: Else we will just treat this key removal as a no-op.
-        // Not throwing error will minimize the need (thus risk) of reusing exposed update OTP.
-      }
+    return document;
+  }
 
-      didDocument.publicKey = [...publicKeyMap.values()];
-    } else if (patch.action === 'add-service-endpoints') {
-      // Find the service of the given service type.
-      let service = undefined;
-      if (didDocument.service !== undefined) {
-        service = didDocument.service.find(service => service.type === patch.serviceType);
+  /**
+   * Removes public keys from document.
+   */
+  private static removePublicKeys (document: DocumentModel, patch: any): DocumentModel {
+    const publicKeyMap = new Map(document.publicKeys.map(publicKey => [publicKey.id, publicKey]));
+
+    // Loop through all given public key IDs and delete them from the existing public key only if it is not a recovery key.
+    for (let publicKey of patch.publicKeys) {
+      const existingKey = publicKeyMap.get(publicKey);
+
+      if (existingKey !== undefined) {
+        publicKeyMap.delete(publicKey);
       }
+      // NOTE: Else we will just treat this key removal as a no-op.
+      // Not throwing error will minimize the need (thus risk) of reusing exposed update OTP.
+    }
 
-      // If service not found, create a new service element and add it to the property.
-      if (service === undefined) {
-        service = {
-          type: patch.serviceType,
-          serviceEndpoint: {
-            '@context': 'schema.identity.foundation/hub',
-            '@type': 'UserServiceEndpoint',
-            instances: patch.serviceEndpoints
-          }
-        };
-
-        if (didDocument.service === undefined) {
-          didDocument.service = [service];
-        } else {
-          didDocument.service.push(service);
-        }
-      } else {
-        // Else we add to the existing service element.
+    document.publicKeys = [...publicKeyMap.values()];
+
+    return document;
+  }
 
-        const serviceEndpointSet = new Set(service.serviceEndpoint.instances);
+  private static addServiceEndpoints (document: DocumentModel, patch: any): DocumentModel {
+    // Find the service of the given service type.
+    let service = undefined;
+    if (document.service !== undefined) {
+      service = document.service.find(service => service.type === patch.serviceType);
+    }
 
-        // Loop through all given service endpoints and add them if they don't exist already.
-        for (let serviceEndpoint of patch.serviceEndpoints) {
-          if (!serviceEndpointSet.has(serviceEndpoint)) {
-            service.serviceEndpoint.instances.push(serviceEndpoint);
-          }
+    // If service not found, create a new service element and add it to the property.
+    if (service === undefined) {
+      service = {
+        type: patch.serviceType,
+        serviceEndpoint: {
+          '@context': 'schema.identity.foundation/hub',
```

Comment:
Not a valid URI.

Reply:
Thanks for calling out, this whole service endpoint will be revamped completely within the week by issue #455.